### PR TITLE
Full SSR support for firebase auth

### PR DESCRIPTION
**Pull request by mentAl-maZe**
Hi there,

I finally managed to dedicate some time to solving the SSR rate limiting issue.
I implemented a complete session management, which should provide much more stable support for SSR client SDK authentication.

I also took the liberty to revamp the templates.

Let me know what you think.

### CODE REVIEW

**Comment by lupas** [comment]
Sweet - on a glimpse it looks great!

Also thanks for revamping the templates.

I'll try to find time to have a closer look, since right now I do not - I try to look at it and test it properly in the next 1-3 days.

One small comment just related to the structure:
- The `/helpers` folder was so far meant for externally exposed [helper functions](https://firebase.nuxtjs.org/guide/helpers/), not internal helpers.
Should we, to not make it confusing, create two seperate folders? Just so we have all (potentially coming) external helper functions in one seperate folder.

Just a small thing :P But since it's different stuff I like to separate it.

Regarding everything else I'll let you know asap!

Thanks again.

**Comment by mentAl-maZe** [comment]
You are absolutely right, even more so as the `SessionManager` class is written in es6 and is possibly not supported by browsers, because its intended use is for the server only.

I created a `./lib/services` folder and put it in there.

File: package.json [hunk 1]
```diff
@@ -29,7 +29,8 @@
     "commit": "npx git-cz"
   },
   "dependencies": {
-    "consola": "^2.11.3"
+    "consola": "^2.11.3",
+    "promise-throttle": "^1.0.1"
```
[the remark this answers is not in the document]
Comment (mentAl-maZe) [reply package.json]: Oops... This was a relic from an experiment... removed

File: lib/module.js [hunk 2]
```diff
@@ -74,8 +74,8 @@ module.exports = function (moduleOptions) {
     })
 
     if (ssrConfig.serverLogin && credential) {
-      const serverLoginConfig =
-        typeof ssrConfig.serverLogin !== 'object' ? {} : ssrConfig.serverLogin
+      const serverLoginConfig = (options.sessions =
```
[the remark this answers is not in the document]
Comment (mentAl-maZe) [reply lib/module.js]: I removed the intermediate variable altogether as it was just used for the fallback (in `render:routeDone`) anyway

**Comment by mentAl-maZe** [comment]
On a sidenote i currently struggle with a problem on firebase functions which is quite puzzling...
I found this issue on the global nuxt repo:
https://github.com/nuxt/nuxt.js/issues/7202

Possible Solution: We might have to inject the services into Vue/Context/Vuex manually...

I'll open an issue on this repo so we can keep track of this here

PS: thank you for the quick replies :grin: 
Sry it took me a bit to respond

**Comment by lupas** [comment]
Hey @mentAl-maZe 

What you gotta keep in mind is that not everyone is using serverLogin - especially when it comes to checking if your changes work.

I just started testing and first thing happening to me:
Screenshot 2020-04-27 at 21 57 01

My auth config:
```js
        auth: {
          initialize: {
            onAuthStateChangedAction: 'onAuthStateChanged'
          },
          ssr: true
        },
```

So I guess something like this would fix it:
```js
    this.lifetime = sessionConfig ? sessionConfig.sessionLifetime : 0
    this.loginDelay = sessionConfig ? sessionConfig.loginDelay : 50
```

I wonder though... the SessionManager is not really used in ssr: true but serverLogin: false settings. no? 

Can we make sure that this "rather complex" implementation is skipped when a user is using the module in standard mode?

I'm currently a bit worried that this all is getting a bit complex for a serverSide login experimental feature that is most likely not going out of experimental status anytime and will only be used by the 0.01%. So I'd kinda like to keep all the setup for it as separate as possible and not mix it with standard "ssr: true" functionality which worked very well without any of that.

Not saying we shouldn't do it, but I honestly barely know how to maintain half of the module anymore since it's server side login logic. So not saying we shouldn't implement it at all, but I just wanna avoid that this is messing with all the rest which worked perfectly so far. Does that make sense?

**Comment by mentAl-maZe** [comment]
> What you gotta keep in mind is that not everyone is using serverLogin - especially when it comes to checking if your changes work.

Yeah... sry... as stated before i don't have a lot of free time at the moment, this is simply an artifact of my work on a project, although i tested it quite extensively... i only tested for my requirements...

> Can we somehow make sure that this "rather complex" implementation is skipped when a user is using the module in "not experimental serverLogin" mode?

In theory it shouldn't, although it is only used on the server side and then only to get the instance if `ssrLogin` is not used.
I'll add a condition to avoid importing it if ssrLogin is not used.

> I'm currently a bit worried that this all is getting a bit complex for a serverSide login experimental feature that is most likely not going out of experimental status anytime and will only be used by the 0.01%. So I'd kinda like to keep all the setup for it as separate as possible and not mix it with standard "ssr: true" functionality which worked very well without any of that.
>
> Not saying we shouldn't do it, but I honestly barely know how to maintain half of the module anymore since it's server side login logic. So not saying we shouldn't implement it at all, but I just wanna avoid that this is messing with all the rest which worked perfectly so far. Does that make sense?

Yes it does...
i tried to export any additional logic (where possible... hooks have to be done in the module) into the service class
The updates to the templates are mainly cosmetic (as it helped with debugging when the resulting plugins were formatted and avoided unnecessary code) and should not interfere with current functionality.

That said, if you want i could migrate the whole ssr thing into a separate package (to be installed as a @nuxtjs/firebase plugin)?

**Comment by mentAl-maZe** [comment]
I gave the templates the old once over to ensure we are using the correct types.
e.g.
if: `options.services.auth = true`
this: `options.services.auth.presistence`
would throw theoretically (i guess errors are suppressed in templates)

**Comment by mentAl-maZe** [comment]
I realize that i have caused quite some headache with the changes to the templates.
I have prepared a branch of off my pull request implementing changes proposed in #165

I'm terribly sorry for the inconvenience... :disappointed: 

If you want i can create a pull request for that branch once this one has been merged, although I'm not quite clear on his implementation of the payload object...

**Comment by lupas** [comment]
> Yeah... sry... as stated before i don't have a lot of free time at the moment, this is simply an artifact of my work on a project, although i tested it quite extensively... i only tested for my requirements...

No worries at all, totally understand. :) Just meant to say that a quick regression test would be appreciated. But I know how it is, happens to me too.

> I'll add a condition to avoid importing it if ssrLogin is not used.

Great thanks!

> The updates to the templates are mainly cosmetic (as it helped with debugging when the resulting plugins were formatted and avoided unnecessary code) and should not interfere with current functionality.

I don't have an issue with these, I appreciate such changes :)

> That said, if you want i could migrate the whole ssr thing into a separate package (to be installed as a @nuxtjs/firebase plugin)?

I actually think that would not be too bad of an idea! You could host that plugin and focus on the server-side SSR issues, while the module can focus on the core.

I would certainly link to that repo as an official plugin - or we can also keep it as a plugin within this repo. But separated from the core code.

> I realize that i have caused quite some headache with the changes to the templates.
I have prepared a branch off of my pull request implementing changes proposed in #165 . I'm terribly sorry for the inconvenience... 😞

No issues at all. I haven't had time to look at #165 properly but it seems like there will be a few changes needed anyway. So I'd like to finish off this PR first. No inconvenience at all.

As mentioned in the other issue, hit me up on Discord so we can talk all this through, I think that will be easier then here.

Thanks for all the support!

**Pull request merged**